README.md
# Firebase Token Generator - Python

[Firebase Custom Login](https://www.firebase.com/docs/web/guide/simple-login/custom.html)
gives you complete control over user authentication by allowing you to authenticate users
with secure JSON Web Tokens (JWTs). The auth payload stored in those tokens is available
for use in your Firebase [security rules](https://www.firebase.com/docs/security/api/rule/).
This is a token generator library for Python which allows you to easily create those JWTs.


## Installation

The Firebase Python token generator library is available via pip:

```bash
$ pip install firebase-token-generator
```

## A Note About Security

**IMPORTANT:** Because token generation requires your Firebase Secret, you should only generate
tokens on *trusted servers*. Never embed your Firebase Secret directly into your application and
never share your Firebase Secret with a connected client.


## Generating Tokens

To generate tokens, you'll need your Firebase Secret which you can find by entering your Firebase
URL into a browser and clicking the "Secrets" tab on the left-hand navigation menu.

Once you've downloaded the library and grabbed your Firebase Secret, you can generate a token with
this snippet of Python code:

```python
from firebase_token_generator import create_token

auth_payload = {"uid": "1", "auth_data": "foo", "other_auth_data": "bar"}
token = create_token("<YOUR_FIREBASE_SECRET>", auth_payload)
```

The payload passed to `create_token()` is made available for use within your
security rules via the [`auth` variable](https://www.firebase.com/docs/security/api/rule/auth.html).
This is how you pass trusted authentication details (e.g. the client's user ID)
to your Firebase security rules. The payload can contain any data of your
choosing, however it must contain a "uid" key, which must be a string of less
than 256 characters. The generated token must be less than 1024 characters in
total.


## Token Options

A second `options` argument can be passed to `create_token()` to modify how Firebase treats the
token. Available options are:

* **expires** (int or datetime) - A timestamp (as number of seconds since the epoch) or `datetime`
denoting the time after which this token should no longer be valid.

* **notBefore** (int or datetime) - A timestamp (as number of seconds since the epoch) or `datetime`
denoting the time before which this token should be rejected by the server.

* **admin** (bool) - Set to `True` if you want to disable all security rules for this client. This
will provide the client with read and write access to your entire Firebase.

* **debug** (bool) - Set to `True` to enable debug output from your security rules. You should
generally *not* leave this set to `True` in production (as it slows down the rules implementation
and gives your users visibility into your rules), but it can be helpful for debugging.

* **simulate** (bool) - If `True`, Firebase will run security rules but not actually make any
data changes. Note that this is internal-only for now.

Here is an example of how to use the second `options` argument:

```python
from firebase_token_generator import create_token

auth_payload = {"uid": "1", "auth_data": "foo", "other_auth_data": "bar"}
options = {"admin": True}
token = create_token("<YOUR_FIREBASE_SECRET>", auth_payload, options)
```


CHANGES.txt
v2.0.1, 2014/12/14 -- Add support for Python 3 
v2.0.0, 2014/09/15 -- Update token generator for updated token format, and require 'uid' field.
v1.3.2, 2013/09/16 -- Support native datetime objects for 'expires' and 'notBefore' options.
v1.0.0, 2013/01/24 -- Initial Release


firebase_token_generator_test.py
try:
    basestring
except NameError:  # Python 3
    basestring = str
from firebase_token_generator import create_token
import unittest

class TestTokenGenerator(unittest.TestCase):

    def test_smoke_test(self):
        token = create_token("barfoo", {"uid": "foo"})
        self.assertIsInstance(token, basestring)

    def test_malformed_key(self):
        with self.assertRaises(ValueError):
            token = create_token(1234567890, {"uid": "foo"})

    def test_no_uid(self):
        with self.assertRaises(ValueError):
            token = create_token("barfoo", {"blah": 5})

    def test_invalid_uid(self):
        with self.assertRaises(ValueError):
            token = create_token("barfoo", {"uid": 5, "blah": 5})

    def test_uid_max_length(self):
        #length:                                        10        20        30        40        50        60        70        80        90       100       110       120       130       140       150       160       170       180       190       200       210       220       230       240       250   256
        token = create_token("barfoo", {"uid": "1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456"})
        self.assertIsInstance(token, basestring)

    def test_uid_too_long(self):
        with self.assertRaises(ValueError):
            #length:                                        10        20        30        40        50        60        70        80        90       100       110       120       130       140       150       160       170       180       190       200       210       220       230       240       250    257
            token = create_token("barfoo", {"uid": "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567"})

    def test_uid_min_length(self):
        token = create_token("barfoo", {"uid": ""})
        self.assertIsInstance(token, basestring)

    def test_token_too_long(self):
        with self.assertRaises(RuntimeError):
            create_token("barfoo", {"uid": "blah", "long_var": "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345612345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234561234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456"})

    def test_no_uid_with_admin(self):
        token = create_token("barfoo", None, {"admin": True})
        self.assertIsInstance(token, basestring)
        token = create_token("barfoo", {}, {"admin": True})
        self.assertIsInstance(token, basestring)
        token = create_token("barfoo", {"foo": "bar"}, {"admin": True})
        self.assertIsInstance(token, basestring)

    def test_invalid_uid_with_admin(self):
        with self.assertRaises(ValueError):
            token = create_token("barfoo", {"uid": 1}, {"admin": True})
        with self.assertRaises(ValueError):
            token = create_token("barfoo", {"uid": None}, {"admin": True})
        with self.assertRaises(ValueError):
            token = create_token("barfoo", "foo", {"admin": True})

if __name__ == '__main__':
    unittest.main()


firebase_token_generator.py
try:
    basestring
except NameError:  # Python 3
    basestring = str
from array import array
from base64 import urlsafe_b64encode
import hashlib
import hmac
import sys
try:
    import json
except ImportError:
    import simplejson as json
import calendar
import time
import datetime

__all__ = ['create_token']

TOKEN_VERSION = 0
TOKEN_SEP = '.'

CLAIMS_MAP = {
    'expires': 'exp',
    'notBefore': 'nbf',
    'admin': 'admin',
    'debug': 'debug',
    'simulate': 'simulate'
}

def create_token(secret, data, options=None):
    """
    Generates a secure authentication token.
 
    Our token format follows the JSON Web Token (JWT) standard:
    header.claims.signature
  
    Where:
    1) "header" is a stringified, base64-encoded JSON object containing version and algorithm information.
    2) "claims" is a stringified, base64-encoded JSON object containing a set of claims:
      Library-generated claims:
      "iat" -> The issued at time in seconds since the epoch as a number
      "d" -> The arbitrary JSON object supplied by the user.
      User-supplied claims (these are all optional):
      "exp" (optional) -> The expiration time of this token, as a number of seconds since the epoch.
      "nbf" (optional) -> The "not before" time before which the token should be rejected (seconds since the epoch)
      "admin" (optional) -> If set to true, this client will bypass all security rules (use this to authenticate servers)
      "debug" (optional) -> "set to true to make this client receive debug information about security rule execution.
      "simulate" (optional, internal-only for now) -> Set to true to neuter all API operations (listens / puts
                 will run security rules but not actually write or return data).
    3) A signature that proves the validity of this token (see: http://tools.ietf.org/html/draft-ietf-jose-json-web-signature-07)
  
    For base64-encoding we use URL-safe base64 encoding. This ensures that the entire token is URL-safe
    and could, for instance, be placed as a query argument without any encoding (and this is what the JWT spec requires).
  
    Args:
        secret - the Firebase Application secret
        data - a json serializable object of data to be included in the token
        options - An optional dictionary of additional claims for the token. Possible keys include:
            a) "expires" -- A datetime or timestamp (as a number of seconds since the epoch) denoting a time after
                            which this token should no longer be valid.
            b) "notBefore" -- A datetime or timestamp (as a number of seconds since the epoch) denoting a time before
                            which this token should be rejected by the server.
            c) "admin" -- Set to true to bypass all security rules (use this for your trusted servers).
            d) "debug" -- Set to true to enable debug mode (so you can see the results of Rules API operations)
            e) "simulate" -- (internal-only for now) Set to true to neuter all API operations (listens / puts
                            will run security rules but not actually write or return data)
    Returns:
        A signed Firebase Authentication Token
    Raises:
        ValueError: if an invalid key is specified in options

    """
    if not isinstance(secret, basestring):
        raise ValueError("firebase_token_generator.create_token: secret must be a string.")
    if not options and not data:
        raise ValueError("firebase_token_generator.create_token: data is empty and no options are set.  This token will have no effect on Firebase.");
    if not options:
        options = {}
    is_admin_token = ('admin' in options and options['admin'] == True)
    _validate_data(data, is_admin_token)
    claims = _create_options_claims(options)
    claims['v'] = TOKEN_VERSION
    claims['iat'] = int(time.time())
    claims['d'] = data

    token = _encode_token(secret, claims)
    if len(token) > 1024:
        raise RuntimeError("firebase_token_generator.create_token: generated token is too long.")
    return token

def _validate_data(data, is_admin_token):
    if data is not None and not isinstance(data, dict):
        raise ValueError("firebase_token_generator.create_token: data must be a dictionary")
    contains_uid = (data is not None and 'uid' in data)
    if ((not contains_uid and not is_admin_token) or (contains_uid and not isinstance(data['uid'], basestring))):
        raise ValueError("firebase_token_generator.create_token: data must contain a \"uid\" key that must be a string.")
    if (contains_uid and (len(data['uid']) > 256)):
        raise ValueError("firebase_token_generator.create_token: data must contain a \"uid\" key that must not be longer than 256 bytes.")

def _create_options_claims(opts):
    claims = {}
    for k in opts:
        if (isinstance(opts[k], datetime.datetime)):
            opts[k] = int(calendar.timegm(opts[k].utctimetuple()))
        if k in CLAIMS_MAP:
            claims[CLAIMS_MAP[k]] = opts[k]
        else:
            raise ValueError('Unrecognized Option: %s' % k)
    return claims

if sys.version_info < (2, 7):
    def _encode(bytes_data):
        # Python 2.6 has problems with bytearrays in b64
        encoded = urlsafe_b64encode(bytes(bytes_data))
        return encoded.decode('utf-8').replace('=', '')
else:
    def _encode(bytes):
        encoded = urlsafe_b64encode(bytes)
        return encoded.decode('utf-8').replace('=', '')


def _encode_json(obj):
    return _encode(bytearray(json.dumps(obj, separators=(',',':')), 'utf-8'))

def _sign(secret, to_sign):
    def portable_bytes(s):
        try:
            return bytes(s, 'utf-8')
        except TypeError:
            return bytes(s)
    return _encode(hmac.new(portable_bytes(secret), portable_bytes(to_sign), hashlib.sha256).digest())

def _encode_token(secret, claims):
    encoded_header = _encode_json({'typ': 'JWT', 'alg': 'HS256'})
    encoded_claims = _encode_json(claims)
    secure_bits = '%s%s%s' % (encoded_header, TOKEN_SEP, encoded_claims)
    sig = _sign(secret, secure_bits)
    return '%s%s%s' % (secure_bits, TOKEN_SEP, sig)
